pkg/check/table_structure.go

```diff
@@ -77,73 +149,249 @@ func (c *TablesChecker) Name() string {
+func (c *TablesChecker) checkAST(stmt ast.StmtNode) []*incompatibilityOption {
 	st, ok := stmt.(*ast.CreateTableStmt)
 	if !ok {
-		return errors.Errorf("Expect CreateTableStmt but got %T", stmt)
+		return []*incompatibilityOption{
+			{
+				state:      StateFailure,
+				errMessage: fmt.Sprintf("Expect CreateTableStmt but got %T", stmt),
+			},
+		}
 	}
-	var err error
 
+	var options []*incompatibilityOption
 	// check columns
 	for _, def := range st.Cols {
-		err = c.checkColumnDef(def)
-		if err != nil {
-			return errors.Trace(err)
+		option := c.checkColumnDef(def)
+		if option != nil {
+			options = append(options, option)
 		}
 	}
 	// check constrains
 	for _, cst := range st.Constraints {
-		err = c.checkConstraint(cst)
-		if err != nil {
-			return errors.Trace(err)
+		option := c.checkConstraint(cst)
+		if option != nil {
+			options = append(options, option)
 		}
 	}
 
 	// check options
 	for _, opt := range st.Options {
-		err = c.checkTableOption(opt)
-		if err != nil {
-			return errors.Trace(err)
+		option := c.checkTableOption(opt)
+		if option != nil {
+			options = append(options, option)
 		}
 	}
-	return nil
+	return options
 }
 
-func (c *TablesChecker) checkColumnDef(def *ast.ColumnDef) error {
+func (c *TablesChecker) checkColumnDef(def *ast.ColumnDef) *incompatibilityOption {
 	return nil
 }
 
-func (c *TablesChecker) checkConstraint(cst *ast.Constraint) error {
+func (c *TablesChecker) checkConstraint(cst *ast.Constraint) *incompatibilityOption {
 	switch cst.Tp {
 	case ast.ConstraintForeignKey:
-		c.warnings++
-		return errors.Errorf("Foreign Key is parsed but ignored by TiDB.")
+		return &incompatibilityOption{
+			state:       StateWarning,
+			instruction: fmt.Sprintf("please ref document: https://github.com/pingcap/docs-cn/blob/master/sql/ddl.md"),
+			errMessage:  fmt.Sprintf("Foreign Key %s is parsed but ignored by TiDB.", cst.Name),
+		}
 	}
+
 	return nil
 }
 
-func (c *TablesChecker) checkTableOption(opt *ast.TableOption) error {
+func (c *TablesChecker) checkTableOption(opt *ast.TableOption) *incompatibilityOption {
 	switch opt.Tp {
 	case ast.TableOptionCharset:
 		// Check charset
 		cs := strings.ToLower(opt.StrValue)
 		if cs != "binary" && !charset.ValidCharsetAndCollation(cs, "") {
-			c.errors++
-			return errors.Errorf("Unsupported charset %s", opt.StrValue)
+			return &incompatibilityOption{
+				state:       StateFailure,
+				instruction: fmt.Sprintf("please ref document: https://github.com/pingcap/docs-cn/blob/master/sql/character-set-support.md"),
+				errMessage:  fmt.Sprintf("unsupport charset %s", opt.StrValue),
+			}
 		}
 	}
 	return nil
 }
+
+// ShardingTablesCheck checks consistency of sharding table structures
+// * check whether they have same column list
+// * check whether they have auto_increment key
+type ShardingTablesCheck struct {
+	dbs    map[string]*sql.DB
+	tables map[string]map[string][]string // instance => {schema: [table1, table2, ...]}
```

Comment:
when I first saw it, I got it wrong. I think `tables` are all tables that need to be check for a whole task. but in fact, they are tables in a sharding group.
add some comments to make it clearer?

and this seems can also be used to check upstream's table and downstream's table?

Reply:
comment for `ShardingTablesCheck`：
`// ShardingTablesCheck checks consistency of table structures of one sharding group`

yep, it can do it implicitly （check upstream's table and downstream's table）